pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -326,13 +327,74 @@ func (c *Cluster) postReconcileUpdateOSDProperties(desiredOSDs map[int]*OSDInfo)
 	for _, actualOSD := range osdUsage.OSDNodes {
-		if desiredOSD, ok := desiredOSDs[actualOSD.ID]; ok {
-			if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
-				// Log the error and allow other updates to continue
-				logger.Error(err)
-			}
+		if err := c.resizeOsdCrushWeight(actualOSD); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Errorf("failed to resize osd crush weight on cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
+		}
+		desiredOSD, ok := desiredOSDs[actualOSD.ID]
+		if !ok {
+			continue
+		}
+		if err := c.updateDeviceClassIfChanged(actualOSD.ID, desiredOSD.DeviceClass, actualOSD.DeviceClass); err != nil {
+			// Log the error and allow other updates to continue
+			logger.Errorf("failed to update device class on cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
 		}
 	}
+
+	if err := c.resizeCrushTreeBucketWeight(); err != nil {
+		// Log the error and allow other updates to continue
+		logger.Errorf("failed to resize crush tree bucket weight for cluster in namespace %s: %v", c.clusterInfo.Namespace, err)
+	}
+	return nil
+}
+
+func (c *Cluster) resizeCrushTreeBucketWeight() error {
+	if !c.spec.Storage.AllowOsdCrushWeightUpdate {
+		return nil
+	}
+	logger.Info("update crush tree bucket weight, as osd are resized for cluster in namespace %q", c.clusterInfo.Namespace)
+	args := []string{"osd", "crush", "reweight-all"}
+	buf, err := cephclient.NewCephCommand(c.context, c.clusterInfo, args).Run()
+	if err != nil {
+		return errors.Wrapf(err, "failed to reweight crush tree bucket weight for cluster in namespace %q: %s", c.clusterInfo.Namespace, string(buf))
+	}
+	return nil
+}
+
+func convertKibibytesToTebibytes(kib string) (float64, error) {
+	kibFloat, err := strconv.ParseFloat(kib, 64)
+	if err != nil {
+		return float64(0), errors.Wrap(err, "failed to convert string to float")
+	}
+	return kibFloat / float64(1024*1024*1024), nil
+}
+
+func (c *Cluster) resizeOsdCrushWeight(actualOSD cephclient.OSDNodeUsage) error {
+	if !c.spec.Storage.AllowOsdCrushWeightUpdate {
+		return nil
+	}
+	crushWeight, err := strconv.ParseFloat(actualOSD.CrushWeight.String(), 64)
+	if err != nil {
+		return errors.Wrapf(err, "failed converting string to float for osd.%d crush weight %q", actualOSD.ID, actualOSD.CrushWeight.String())
+	}
+	// actualOSD.KB is in KiB units
+	desiredOSDTiB, err := convertKibibytesToTebibytes(actualOSD.KB.String())
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert KiB to TiB for osd.%d crush weight %q", actualOSD.ID, actualOSD.KB.String())
+	}
+	// do not reweight if the desired crush weight is 0 or they are equal or there percentage resize is less than 5 percent
+	if desiredOSDTiB == float64(0) || desiredOSDTiB == crushWeight || math.Abs(((desiredOSDTiB-crushWeight)/crushWeight)) < 0.05 {
```

Comment:
Can we restructure this a bit for more clear logging? Also, if we still have a calculation discrepancy compared to the ceph discrepancy, let's up the diff to 10%. I'm imagining if users are increasing an OSD size, it will be by at least 10%, and most commonly at least 50%.
```suggestion
if desiredOSDTiB == float64(0) {
 // debug log message about weight being 0
 return nil
} else if math.Abs(((desiredOSDTiB-crushWeight)/crushWeight)) < 0.1 {
 // debug log message about weight diff being too small to update
 return nil
}
```